File: arduino.js
const SerialPort = require('serialport');
const Readline = require('@serialport/parser-readline');
const port = new SerialPort('/dev/ttyACM0', { baudRate: 115200 });
const parser = port.pipe(new Readline({ delimiter: '\n' }));

port.on("open", () => {
    console.log('serial port open');
});

parser.on('data', data => {
    console.log('got word from arduino:', data);
});

function write(data){
    port.write(data);
}

let writting = false
function writeUnreliable(data){
    if (writting) return
    writting = true
    port.write(data, () => {
        writting = false
    });
}

module.exports = {
    keyClick: (key) => write(`k${key}\n`),
    keyPress: (key) => write(`u${key}\n`),
    keyRelease: (key) => write(`y${key}\n`),
    mouseMove: (x, y) => writeUnreliable(`m${x}${y}\n`),
    mouseMoveRelative: (x, y) => writeUnreliable(`z${x}${y}\n`),
    mousePress: () => write(`p\n`),
    mouseRightClick: () => write(`b\n`),
    mouseRelease: () => write(`h\n`),
    mouseDoubleClick: async() => {
        this.mousePress()
        await new Promise(r => setTimeout(r, 500));
        this.mouseRelease()
        await new Promise(r => setTimeout(r, 500));
        this.mousePress()
        await new Promise(r => setTimeout(r, 500));
        this.mouseRelease()
    },
}

File: arduino.py
# Importing Libraries
import serial
import time
arduino = serial.Serial(port='COM5', baudrate=115200, timeout=.1)
arduino.timeout = None


def keyclick(key: str):
    arduino.write(bytes(f"k{key}\n", 'utf-8'))


def keypress(key: str):
    arduino.write(bytes(f"u{key}\n", 'utf-8'))


def keyrelease(key: str):
    arduino.write(bytes(f"y{key}\n", 'utf-8'))


def movemouse(x: int, y: int):
    arduino.write(bytes(f"m{str(x).zfill(4)}{str(y).zfill(4)}\n", 'utf-8'))
    print(arduino.readline())


def movemouserelative(x: int):
    arduino.write(bytes(f"z{str(x).zfill(4)}\n", 'utf-8'))


def mousepress():
    arduino.write(bytes(f"p\n", 'utf-8'))


def mouserightclick():
    arduino.write(bytes(f"b\n", 'utf-8'))


def mouserelease():
    arduino.write(bytes(f"h\n", 'utf-8'))


def mouseclick():
    mousepress()
    time.sleep(.5)
    mouserelease()
